nifi-extension-bundles/nifi-box-bundle/nifi-box-processors/src/main/java/org/apache/nifi/processors/box/UpdateBoxFileMetadataInstance.java

```
@@ -248,40 +207,110 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
+            final String propertyPath = "/" + fieldName;
+
+            if (updateField(metadata, propertyPath, value, currentKeys.contains(propertyPath))) {
+                processedKeys.add(fieldName);
+            }
+        }
+
+        return processedKeys;
+    }
+
+    private boolean updateField(final Metadata metadata,
+                                final String propertyPath,
+                                final Object value,
+                                final boolean exists) {
+        if (value == null) {
+            return false;
         }
 
-        final Object valueObj = valueValues.getFirst().getValue();
-        final String value = valueObj != null ? valueObj.toString() : null;
+        if (exists) {
+            final Object currentValue = metadata.getValue(propertyPath);
 
-        // Add the key-value pair to the metadata update
-        metadata.add("/" + key, value);
-        updatedKeys.add(key);
+            // Only update if values are different
+            if (Objects.equals(currentValue, value)) {
+                return false;
+            }
+
+            // Update
+            switch (value) {
+                case Number n -> metadata.replace(propertyPath, n.doubleValue());
+                case List<?> l -> metadata.replace(propertyPath, convertListToStringList(l));
+                default -> metadata.replace(propertyPath, value.toString());
+            }
+        } else {
+            // Add new field
+            switch (value) {
+                case Number n -> metadata.add(propertyPath, n.doubleValue());
+                case List<?> l -> metadata.add(propertyPath, convertListToStringList(l));
+                default -> metadata.add(propertyPath, value.toString());
+            }
+        }
+        return true;
+    }
+
+    private List<String> convertListToStringList(final List<?> list) {
+        // Box doesn't support null values in metadata lists, so we filter them out
```

Does it make sense to fail instead of silently ignoring null values?

---

Yeah, fails now